gwyddion: 2.48 -> 2.55 (and add options)

cgevans opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Gwyddion has a large number of optional dependencies. In this update, those dependencies have been added with options set mostly to true, and an option (default false) for working Python 2.7 support in the program and in Python (but without the gwyutils module) has been added. I have also added myself as a maintainer as no maintainer is currently set.

Upstream has made the gwyutils module intentionally hard to use by placing it outside the Python path, and I'm not sure the best way to fix this, or whether it should be fixed.  I have also not added options for ruby and perl support, as I don't know how that support works (it isn't listed in the configuration summary).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

-  [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)

(Tested both gwyddion (and confirmed no Python support) and python27Packages.gwyddion (and confirmed Python support))

- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"` 
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cgevans commented:
@P-E-Meunier, it looks like by some coincidence you were actually the original author of this derivation?

P-E-Meunier commented:
Yes, when I noticed you hadn't done it yet ;-)

P-E-Meunier commented:
Adding an install phase takes care of your problem with the python path:                  ~
```                                                                                  
  installPhase = ''
    make install   
    mkdir -p $out/${pythonPackages.python.sitePackages}
    cp $out/share/gwyddion/pygwy/* $out/${pythonPackages.python.sitePackages}
  '';                                                                        
```  

But I can't get the python module to work, I keep getting an error: "could not import gtk". I don't understand where the python files are installed, do you?

cgevans commented:
>But I can't get the python module to work, I keep getting an error: "could not import gtk". I don't understand where the python files are installed, do you?

That sounds like a dependency problem with its dependency on pygtk.  It works for me, but I need to use something like the following: `python27.withPackages (ps: [ps.gwyddion])`.  It may be that gwyddion's mechanism of interfacing with python is somehow escaping Nix's paths, and so works fine if Python 2.7 on the current system has pygtk/etc, but not if python27Packages.gwyddion is installed separately?

I'm primarily testing this in Python, not in the embedded console, as my primary use-case is scripting for consistent batch levelling and conversion, but it looks like the embedded console is also working for me.

P-E-Meunier commented:
> works fine if Python 2.7 on the current system has pygtk/etc, but not if python27Packages.gwyddion is installed separately?

That is what `propagatedBuildInputs` are for. Btw, I can `import pygtk` fine, just not "gtk".

jonringer commented on maintainers/maintainer-list.nix:
```diff
@@ -1257,6 +1257,20 @@
+        fingerprint = "669C 1D24 5A87 DB34 6BE4  3216 1A1D 58B8 6AE2 AABD";
+      }
+    ];
+  };
```
should be a separate commit:

```
maintainers: add cge
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
+assert openglSupport -> gnome2 != null;
+assert pythonSupport -> (pythonPackages != null && gnome2 != null);
+
+let version = "2.55";
```
```suggestion
let
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
-let version = "2.48"; in
 stdenv.mkDerivation {
   pname = "gwyddion";
   inherit version;
```
```suggestion
   version = "2.55";
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
+in
 
-let version = "2.48"; in
 stdenv.mkDerivation {
```
```suggestion
stdenv.mkDerivation rec {
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
-    sha256 = "119iw58ac2wn4cas6js8m7r1n4gmmkga6b1y711xzcyjp9hshgwx";
+    url = "mirror://sourceforge/gwyddion/gwyddion-${version}.tar.xz";
+    sha256 = "0l00zszvginpriv12idc0y1x28qmicdmrwkqa007srkxvrdgxwdi";
   };
```
please add some space between fields
```suggestion
  };
  
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
   };
-  nativeBuildInputs = [ pkgconfig ];
-  buildInputs = [ gtk2 ];
+  nativeBuildInputs = [ pkgconfig file ];
```
please use canonical pkg-config name, instead of alias
```suggestion
  nativeBuildInputs = [ pkg-config file ];
  
```

jonringer commented on pkgs/applications/science/chemistry/gwyddion/default.nix:
```diff
@@ -1,17 +1,67 @@
+    optional libzipSupport libzip;
+
+  propagatedBuildInputs = with stdenv.lib;
+    optionals pythonSupport [pygtk pygobject2 python gnome2.gtksourceview];
```
```suggestion
    optionals pythonSupport [ pygtk pygobject2 python gnome2.gtksourceview ];
```

cgevans committed:
maintainers: add cge

cgevans committed:
gwyddion: 2.48 -> 2.55 (and add options)

Gwyddion has a large number of optional dependencies. In this update,
those dependencies have been added with options set mostly to true,
and an option (default false) for working Python 2.7 support in the
program and in Python has been added. I have also added myself as a
maintainer as no maintainer is currently set.

jonringer reviewed (commented):
please squash the fixup commits:
```
git reset HEAD^^
git add pkgs
git commit --amend --no-edit
git push ... ... --force
```

jonringer commented:
https://nixos.org/nixpkgs/manual/#submitting-changes-making-patches

cgevans commented:
Done.  Wasn't sure whether I should squash immediately, or leave the fixups there to show them for review before squashing.

cgevans commented:
@P-E-Meunier, does this work for you now, at least with `python27.withPackages (ps: [ps.gwyddion])`?  I had another python 2.7 package that was confusing me about propagated dependencies; it should be fixed now.

In theory, this should build on macOS; maybe Damien could test it?  The python API is enormously useful for consistent mass conversion of AFM images.

P-E-Meunier commented:
I does work, and TIL it's actually how Python packages are meant to be used:

https://nixos.org/nixpkgs/manual/#installing-python-and-packages

About OSX, I don't know how pygtk works there.

jonringer commented:
> python27.withPackages (ps: [ps.gwyddion])

`withPackages` will install the packages in the base site-packages, having a list of python modules will just make them appear in PYTHONPATH

jonringer reviewed (approved):
diff LGTM
commits LGTM

```
https://github.com/NixOS/nixpkgs/pull/83114
2 package built:
gwyddion python27Packages.gwyddion
```

congrats to @cgevans on his first PR :)

P-E-Meunier commented:
I didn't know it was @cgevans' first PR. Congrats! Now I can't wait to meet the third user of Gwyddion in Nixpkgs.
